IONFileTrasnsferLib/Delegates/IONFLTRDownloadDelegate.swift

```diff
@@ -0,0 +1,103 @@
+import Foundation
+
+/// A delegate class for handling download tasks in a `URLSession`.
+///
+/// The `IONFLTRDownloadDelegate` class manages the progress, completion, and redirection of download tasks.
+/// It communicates progress, success and failure through a publisher.
+///
+/// - Note: This class conforms to `URLSessionTaskDelegate` to handle task-specific events.
+class IONFLTRDownloadDelegate: NSObject {
+    
+    /// The publisher used to send progress, success and failure updates.
+    private var publisher: IONFLTRPublisher
+    
+    /// The destination URL where the downloaded file will be saved.
+    private var destinationURL: URL
+
+    /// A flag indicating whether HTTP redirects should be disabled.
+    private var disableRedirects: Bool
+
+    /// The total number of bytes written during the download.
+    private var totalBytesWritten: Int64 = 0
+
+    /// Initializes a new instance of the `IONFLTRDownloadDelegate` class.
+    ///
+    /// - Parameters:
+    ///   - publisher: The publisher used to send progress and success updates.
+    ///   - destinationURL: The URL where the downloaded file will be saved.
+    ///   - disableRedirects: A flag indicating whether HTTP redirects should be disabled. Defaults to `false`.
+    init(publisher: IONFLTRPublisher, destinationURL: URL, disableRedirects: Bool = false) {
+        self.publisher = publisher
+        self.destinationURL = destinationURL
+        self.disableRedirects = disableRedirects
+    }
+}
+
+extension IONFLTRDownloadDelegate: URLSessionTaskDelegate {
+    
+    /// Reports the progress of a download task.
+    ///
+    /// - Parameters:
+    ///   - session: The `URLSession` containing the download task.
+    ///   - downloadTask: The `URLSessionDownloadTask` reporting progress.
+    ///   - bytesWritten: The number of bytes written since the last call.
+    ///   - totalBytesWritten: The total number of bytes written so far.
+    ///   - totalBytesExpectedToWrite: The total number of bytes expected to be written.
+    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
+        self.totalBytesWritten = totalBytesWritten
+        publisher.sendProgress(totalBytesWritten, totalBytesExpected: totalBytesExpectedToWrite)
+    }
+    
+    /// Handles the completion of a download task.
+    ///
+    /// - Parameters:
+    ///   - session: The `URLSession` containing the download task.
+    ///   - downloadTask: The `URLSessionDownloadTask` that finished downloading.
+    ///   - location: The temporary file location of the downloaded file.
+    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
+        do {
+            try FileManager.default.moveItem(at: location, to: destinationURL)
+
+            let response = downloadTask.response as? HTTPURLResponse
```

Comment:
Should we check for the case where `downloadTask.response` returns `nil`?

We could do something like the following, using `guard`:

```
guard let response = downloadTask.response as? HTTPURLResponse else {
    publisher.sendFailure(someExceptionOrError)
    return
}
```

Reply:
Oh ok yeah that's a good point